### PR TITLE
ubuntu/bootstrap: Patch debootstrap to not create devices under /dev

### PR DESCRIPTION
**Pull request by drinkcat**
More proper fix...

### CODE REVIEW

File: installer/ubuntu/bootstrap [hunk 1]
```diff
@@ -54,6 +51,9 @@ else
     echo "Unable to patch debootstrap, moving on..." 1>&2
 fi
 
+# Patch debootstrap so that is does not create devices under /dev (issue #2387).
+sed -i -e 's/\(setup_devices () {\)/\1 return 0/' "$tmp/functions"
```
Comment (dnschneid) [inline installer/ubuntu/bootstrap]: Drop the parens and use `\0` instead?

File: installer/ubuntu/bootstrap [hunk 2]
```diff
@@ -9,24 +9,21 @@
...
+    f="`wget -O- --no-verbose --timeout=60 -t2 "$d" \
+            | sed -ne 's ^.*\(debootstrap_[0-9.]*.tar.gz\).*$ \1 p' \
+            | tail -n 1`"
+    if [ -z "$f" ]; then
+        error 1 'Failed to download debootstrap.
+Check your internet connection or proxy settings and try again.'
+    fi
     v="${f#*_}"
-    v="${v%.tar.gz}"
+    v="${v%.tar.xz}"
```
Comment (dnschneid) [inline installer/ubuntu/bootstrap]: Should this actually be reverted?  This looked like a bug fix to me (a la #1926)

**Comment by dnschneid** [comment]
Maybe the code to create the empty tarball should also be dropped.

File: installer/ubuntu/bootstrap (hunk 1 above)
Comment (drinkcat) [inline installer/ubuntu/bootstrap]: I did that because I could not find a reference to `\0` in the man page, nor POSIX. But, yeah, most probably ok ,-)

**Comment by drinkcat** [comment]
> Maybe the code to create the empty tarball should also be dropped.

I'll double-check but I think it made deboostrap fail, even with the hack.

File: installer/ubuntu/bootstrap (hunk 2 above)
Comment (drinkcat) [inline installer/ubuntu/bootstrap]: Nope, that's right.

**Comment by drinkcat** [comment]
> I'll double-check but I think it made deboostrap fail, even with the patch.

Yes, this (`debootstrap`), at the end of first stage:

```
	if ! am_doing_phase second_stage; then
...
		cp $DEBOOTSTRAP_DIR/devices.tar.gz	 "$TARGET/debootstrap/devices.tar.gz"
...
	fi
```

Now, I mean an empty file would also do the trick...

File: installer/ubuntu/bootstrap (hunk 1 above)
Comment (drinkcat) [inline installer/ubuntu/bootstrap]: Actually, the regex is missing checks for line beginning & end, so it could match another function.

**Comment by drinkcat** [comment]
Alright, fixes in fixup commit.

**Comment by rfc1036** [comment]
I am the author of the debootstrap change that apparently caused this regression. Do I understand correctly that debootstrap is being run in an environment which does not allow creating device nodes at all?

**Comment by drinkcat** [comment]
> I am the author of the debootstrap change that apparently caused this regression.

Nice, thanks for stopping by ,-)

> Do I understand correctly that debootstrap is being run in an environment which does not allow creating device nodes at all?

Actually, we first call deboostrap first stage, then we bind mount `/dev` from outside the chroot (among other things), chroot, then call `deboostrap --second-stage`.

I did try unmounting /dev before running second-stage, but ended up with a "normal" /dev/null file with some content, as if some parts of deboostrap were expecting a valid /dev/null, then attempted to recreate it. I did not dig further into the issue...

At a minimum, it would be good to have an environment variable that could bypass the devices creation.

**Comment by glaubitz** [comment]
The two-staged debootstrap process is also now broken in Debian unstable meaning it's no longer possible to debootstrap a chroot for a foreign architecture.

Previously, I could debootstrap --foreign onto a blank drive, hook that drive up to a single-board computer and let the machine boot with init=/bin/bash, then run ./debootstrap/debootstrap --second-stage.

With 1.0.75, second-stage worked out-of-the-box. With .76 and .77, debootstrap exits with error 1.

**Comment by drinkcat** [comment]
@glaubitz : You should probably file a Debian bug about this...

**Comment by glaubitz** [comment]
I will once I have verified the exact problem. I just noticed that downgrading debootstrap fixes my issue.

**Pull request merged**